### ISSUE TITLE
Support for Guava Immutable collections

### ISSUE THREAD

**Issue by mattiasoamz** [mapstruct/mapstruct#1025]
It would be great if MapStruct could support Google Guava [immutable collections](https://github.com/google/guava/wiki/ImmutableCollectionsExplained). This would require some custom logic for how to create and populate the collections - for Guava collections you would typically call a static `copyOf()` factory method instead of using a no args constructor and `addAll()`.

This related to #73 in the sense that together would support truly immutable object graphs.

Below follows an example of the desired functionality:

```java
public class SourceBean {
    private Map<String, String> dictionaryOfStrings;

    public Map<String, String> getDictionaryOfStrings() {
        return dictionaryOfStrings;
    }

    public void setDictionaryOfStrings(Map<String, String> dictionaryOfStrings) {
        this.dictionaryOfStrings = dictionaryOfStrings;
    }
} 

public class TargetBean {
    private final ImmutableMap<String, String> dictionaryOfStrings;

    public ImmutableMap<String, String> getDictionaryOfStrings() {
        return dictionaryOfStrings;
    }

    public void setDictionaryOfStrings(ImmutableMap<String, String> dictionaryOfStrings) {
        this.dictionaryOfStrings = dictionaryOfStrings;
    }
} 

@Mapper
public interface SomeMapper {
    public TargetBean map(SourceBean src);

    public ImmutableList<TargetBean> map(Collection<SourceBean> src);
}
```

The example would obviously make more sense if the ´TargetBean` class was completely immutable but that would require setting property values through constructors, which is covered in #73, and I don't want to drag that functionality into this issue.

**Comment by gunnarmorling** [comment]
Hey @mattiasoamz, having support for Guava's collections would be great. I don't see any general road block to do so, it could be done in exactly the same way as for the already supported collection types. 

Would you maybe interested in helping out with implementing this? It's quite straight-forward and we can guide you should you run into any problems.

**Comment by mattiasoamz** [comment]
Ok, I can try to find some time if you point me to where that code lives as well as to some conventions on how to handle new dependencies, writing tests, etc.

**Comment by yhrn** [comment]
This is me again but using another account. I'll be using this from now on.

**Comment by gunnarmorling** [comment]
Hey, thanks for your interest!

As a starter, you could take a look at [IterableMappingMethod](https://github.com/mapstruct/mapstruct/blob/master/processor/src/main/java/org/mapstruct/ap/internal/model/IterableMappingMethod.java), `MapMappingMethod`, how they are handled in the processor pipeline and [their templates](https://github.com/mapstruct/mapstruct/blob/master/processor/src/main/resources/org/mapstruct/ap/internal/model/IterableMappingMethod.ftl).

The linking of public type (e.g. `Set`) to their implementations (e.g. `HashSet`) is done in `TypeFactory`, there we'd need similar entries for the Guava types.

We wouldn't need any new dependency, we already use Guava for testing, and the processor itself doesn't have any hard dependencies to the types used in the compilation. For tests, just check out what's there already in the _processor_ module, we've got some infra for writing annotation processor tests using a programmatic compilation.

Also checkout the pointers given [here](http://mapstruct.org/development/technical-documentation/), and finally, @sjaakd, @agudian and @filiphr may have further tips, too. Happy hacking :)

**Comment by yhrn** [comment]
Hey, so I finally got around to looking at this. Unfortunately, I'm struggeling to find the time at the moment. I have got it working for the basic `ImmutableMap` cases but there are some things that aren't obvious to me how it should work.

1. How should mappers with `@MappingTarget` be treated? It could be disallowed for immutable collections (which is what I've done right now) or we could allow passing in a Builder instance (which complicates the implementation a fair bit).
1. How should `@BeforeMapping`/`@AfterMapping` callbacks with `@MappingTarget`? It could be unsupported for immutable collections (which is what I currently have) or we could support the associated Builder type (once again complicating the implementation a fair bit). For the `@AfterMapping` callback it would actually be possible to pass the final collection but that feels a bit asymmetric to me.

Depending on how you feel about the above I could have an initial PR with only ImmutableMap support (I'll add Collection once I've gotten some feedback since that will follow the same pattern) ready very soon or it might take a while.

**Comment by agudian** [comment]
That sounds great! I'd say open the PR and we'll check it out more closely.

Regarding your questions:

1. I'd say it is okay to not try to update the immutable maps in update-mappings 😁. We could in that case just invoke the setter with a freshly created instance.
2. There should be no special handling for immutable types regarding the lifecycle-methods (`@BeforeMapping`/`@AfterMapping`), as they don't only serve the purpose to modify the provided instance - they could also be used for something totally different (logging, wrapping the instance into a delegator, counting the prime numbers in the key-set or whatever...). So I wouldn't expect any change in that area.

If you want, you could just open the PR and we'll check it out before you continue with the collections. Your choice. And don't feel pressured, just work on your own pace. We're all doing this for fun and often have our hands full with other stuff as well - that's why reviews of the more complex stuff might also take a bit longer with us... 😃. Your help is highly appreciated in any case 👍!

**Comment by Maaartinus** [comment]
>  we could support the associated Builder type (once again complicating the implementation a fair bit)

There isn't much an `ImmutableMap.Builder` can do. It basically can only add things, for anything else you'd need an old good `LinkedHashMap` (which can used as a builder and preserves the iteration order, too, but is less efficient).

I guess, that being able to modify things in `@BeforeMapping`/`@AfterMapping` may be useful even for immutables, but by default I'd do nothing as that's both simpler and faster. If such a feature gets requested, then it can IMHO be supported without much effort by working on a mutable counterpart and in the very last step replacing it by `ImmutableMap.copyOf(it)` or alike.

**Comment by yhrn** [comment]
@agudian:

1. My thinking there was that one of the use cases for `@MappingTarget` is that you might want to be able to pass in a subclass.
2. The main problem with the `@BeforeMapping` callback with `@MappingTarget` is that there is no real result object to pass in until after the X.Builder.build() method is called (which is the very last thing). That means I can't see any way it could behave in the same way. For an `@AfterMapping` callback with `@MappingTarget` I guess it might be up to the use case whether you need a mutable thing (either a builder, or a mutable collection/map) or the actual resulting immutable instance.

**Comment by yhrn** [comment]
Thinking a bit more about this I believe that not supporting the features I brought up make sense from the perspective of being able to be consistent with what can be supported for immutable targets in general (see #73). Builders and mutable counterparts exists in the case of Guava Immutable Collections but won't exist for an arbitrary immutable bean where everything needs to be passed in using a constructor.

The feature that I can see working for all immutable objects is `@AfterMapping` with `@MappingTarget` but that could be added later as suggested by @Maaartinus.

**Comment by agudian** [comment]
> The main problem with the @BeforeMapping callback with @MappingTarget is that there is no real result object to pass in until after the X.Builder.build() method is called (which is the very last thing). That means I can't see any way it could behave in the same way.

Ah, right, that wouldn't work... but perhaps if the before-mapping method would receive a builder? But that can wait for a later iteration where we can see if anyone has the need for such a thing.

**Comment by agudian** [comment]
Using constructors for immutable beans is in the works already (the PR is dormant currently, we'll have to iron out some conceptual stuff), a more general support for builders is on the backlog somewhere as well... So this is a great fit into the big picture of supporting immutables on a brad range... 🙂

**Comment by filiphr** [comment]
@yhrn I am not sure if you had a look at the `StreamMappingMethod.ftl`. `Stream` is also an immutable class and we still have the callbacks engaging (not always though). You can have a look there how we are handling the callbacks. 

1. I agree with you here you can just ignore the updates and only do the callbacks. We do that for the `Stream`(s)
1. There are 3 possible callbacks that we can have:
    - `@BeforeMapping` without `@MappingTarget` - Here I agree it makes no sense to use it. 
    - `@BeforeMapping` with `@MappingTarget` - Also doesn't make sense. However, in #469 we allow `@BeforeMapping` and `@AfterMapping` methods to return a value and if the returned value is not `null` then we return that value immediately. Right now I don't see a use case for this, but we allow it for `Stream`. To do that we just pass an empty `Stream` to those methods.
    - `@AfterMapping` I would say that we pass the `Immutable` object and not the builder. If they want to modify it then they need to create a new one and return it again.

**Comment by yhrn** [comment]
@filiphr 
I haven't looked at `StreamMappingMethod.ftl`, I'll do that as soon as I get the chance. Meanwhile, there is one thing in your answer that confuses me. Why wouldn't `@BeforeMapping` without `@MappingTarget` make sense to use? In my mind that would work the same for immutable collections as it does for any other target type since the target is not included in the callback. Am I misunderstanding something here?

**Comment by filiphr** [comment]
@yhrn You are right. I got confused and wrote something wrong. Your understanding is correct.

**Comment by BigMichi1** [comment]
looks like an amazing addition to mapstruct. is there something i can already try out maybe a fork/branch to support developing this feature?

**Comment by filiphr** [comment]
@BigMichi1 if you are interested in working on this then have a look at the comments that were written in this and let us know if the directions are clear enough. Currently no one from the MapStruct team and it seems the community is not working on this.

We are always happy to have contributors to help us with new features.
